pkg/kubelet/pod_workers.go

```diff
@@ -1030,240 +1294,254 @@ func (p *podWorkers) acknowledgeTerminating(pod *v1.Pod) PodStatusFunc {
-		p.workQueue.Enqueue(pod.UID, wait.Jitter(p.backOffPeriod, workerBackOffPeriodJitterFactor))
-	}
-	p.completeWorkQueueNext(pod.UID)
-}
-
-// completeWorkQueueNext holds the lock and either queues the next work item for the worker or
-// clears the working status.
-func (p *podWorkers) completeWorkQueueNext(uid types.UID) {
-	p.podLock.Lock()
-	defer p.podLock.Unlock()
-	if workUpdate, exists := p.lastUndeliveredWorkUpdate[uid]; exists {
-		p.podUpdates[uid] <- workUpdate
-		delete(p.lastUndeliveredWorkUpdate, uid)
-	} else {
-		p.podSyncStatuses[uid].working = false
+		p.workQueue.Enqueue(podUID, wait.Jitter(p.backOffPeriod, workerBackOffPeriodJitterFactor))
 	}
-}
 
-// contextForWorker returns or initializes the appropriate context for a known
-// worker. If the current context is expired, it is reset. If no worker is
-// present, no context is returned.
-func (p *podWorkers) contextForWorker(uid types.UID) context.Context {
+	// if there is a pending update for this worker, requeue immediately, otherwise
+	// clear working status
 	p.podLock.Lock()
 	defer p.podLock.Unlock()
-
-	status, ok := p.podSyncStatuses[uid]
-	if !ok {
-		return nil
-	}
-	if status.ctx == nil || status.ctx.Err() == context.Canceled {
-		status.ctx, status.cancelFn = context.WithCancel(context.Background())
+	if status, ok := p.podSyncStatuses[podUID]; ok {
+		if status.pendingUpdate != nil {
+			select {
+			case p.podUpdates[podUID] <- struct{}{}:
+				klog.V(4).InfoS("Requeueing pod due to pending update", "podUID", podUID)
+			default:
+				klog.V(4).InfoS("Pending update already queued", "podUID", podUID)
+			}
+		} else {
+			status.working = false
+		}
 	}
-	return status.ctx
 }
 
 // SyncKnownPods will purge any fully terminated pods that are not in the desiredPods
 // list, which means SyncKnownPods must be called in a threadsafe manner from calls
-// to UpdatePods for new pods. It returns a map of known workers that are not finished
-// with a value of SyncPodTerminated, SyncPodKill, or SyncPodSync depending on whether
-// the pod is terminated, terminating, or syncing.
-func (p *podWorkers) SyncKnownPods(desiredPods []*v1.Pod) map[types.UID]PodWorkerState {
+// to UpdatePods for new pods. Because the podworker is dependent on UpdatePod being
+// invoked to drive a pod's state machine, if a pod is missing in the desired list the
+// pod worker must be responsible for delivering that update. The method returns a map
+// of known workers that are not finished with a value of SyncPodTerminated,
+// SyncPodKill, or SyncPodSync depending on whether the pod is terminated, terminating,
+// or syncing.
+func (p *podWorkers) SyncKnownPods(desiredPods []*v1.Pod) (map[types.UID]PodWorkerState, map[types.UID]struct{}) {
 	workers := make(map[types.UID]PodWorkerState)
 	known := make(map[types.UID]struct{})
 	for _, pod := range desiredPods {
 		known[pod.UID] = struct{}{}
 	}
+	var restartable map[types.UID]struct{}
```

Comment:
nit: why not also initalize the map here with make, to avoid having to check if it's nil in both places below?